File: src/main/java/io/github/svegon/capi/event/render/tooltip/ShowTooltipSectionCallback.java
package io.github.svegon.capi.event.render.tooltip;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.item.ItemStack;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

@FunctionalInterface
public interface ShowTooltipSectionCallback {
    Event<ShowTooltipSectionCallback> EVENT = EventFactory.createArrayBacked(ShowTooltipSectionCallback.class,
            (flags, tooltipSection, callback) -> {}, listeners -> (flags, tooltipSection, callback) -> {
        for (ShowTooltipSectionCallback listener : listeners) {
            listener.shouldShowTooltipSection(flags, tooltipSection, callback);

            if (callback.isCancelled()) {
                return;
            }
        }
    });

    void shouldShowTooltipSection(int flags, ItemStack.TooltipSection tooltipSection,
                                  CallbackInfoReturnable<Boolean> callback);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/render/tooltip/ShowTooltipSectionCallback.kt
package io.github.svegon.capi.event.render.tooltip

import io.github.svegon.capi.event.render.tooltip.ShowTooltipSectionCallback
import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.item.ItemStack
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable

fun interface ShowTooltipSectionCallback {
    fun shouldShowTooltipSection(
        flags: Int, tooltipSection: ItemStack.TooltipSection?,
        callback: CallbackInfoReturnable<Boolean?>?
    )

    companion object {
        @JvmField
        val EVENT: Event<ShowTooltipSectionCallback> = EventFactory.createArrayBacked<ShowTooltipSectionCallback>(
            ShowTooltipSectionCallback::class.java,
            ShowTooltipSectionCallback { flags: Int, tooltipSection: ItemStack.TooltipSection?, callback: CallbackInfoReturnable<Boolean?>? -> }) { listeners: Array<ShowTooltipSectionCallback> ->
            ShowTooltipSectionCallback { flags: Int, tooltipSection: ItemStack.TooltipSection?, callback: CallbackInfoReturnable<Boolean?> ->
                for (listener in listeners) {
                    listener.shouldShowTooltipSection(flags, tooltipSection, callback)

                    if (callback.isCancelled) {
                        return@ShowTooltipSectionCallback
                    }
                }
            }
        }
    }
}
